optuna/serialization.py

```diff
@@ -0,0 +1,20 @@
+import pickle
+
+import optuna
+
+
+def save_study(study: optuna.study.Study, path: str) -> None:
+    study_in_memory = optuna.create_study(directions=study.directions)
+    for trial in study.trials:
+        study_in_memory.add_trial(trial)
+
+    with open(path, "wb") as f:
+        pickle.dump(study_in_memory, f)
+
+
+def load_study(study: optuna.study.Study, path: str) -> None:
```

Comment:
[RFC] An alternative interface could be as follows (I'm not sure which one is better though):
```python
def load_study(path: str, *create_study_args, **create_study_kwargs) -> optuna.study.Study:
    with open(path, "rb") as f:
        study_in_memory = pickle.load(f)
        
    study = optuna.create_study(*create_study_args, directions=study_in_memory.directions, **create_study_kwargs)
    ...
```

Reply:
Another topic is that the name `load_study` sounds confusing as there's already a method that has the same name (`optuna.study.load_study`).

Reply:
Thanks for the input. I was thinking about something like your alternative as well (even placing it next to `optuna.create_study`, `optuna.load_study`, an additional for instance `optuna.import_study` to somewhat align parameters to the former two, but with an additional path argument). I'm wondering though if it's not cleaner trying to symmetrize the interfaces between save and load and keep it separated from the study construction functions. I was imagining raising errors in case there are incompatibilities such as the number of objectives or mismatch in directions. What do you think?

As for the conflicting name of `load_study`, I agree that it can be improved. We could simply rename it to `import_study` (along with `export_study`). We could also make those functions `Study` methods `Study.{save,load}`.

Reply:
Thank you for your response.

>  to symmetrize the interfaces between save and load and keep it separated from the study construction functions.  I was imagining raising errors in case there are incompatibilities such as the number of objectives or mismatch in directions.

Sounds good!

> We could also make those functions Study methods `Study.{save,load}`.

Personally, I like this option (but want to hear opinions from other developers).

Reply:
Hm, I'm also beginning to think that the user loading the study would want to do that without prior knowledge about the study. More inputs from others would be fruitful.

Reply:
>  without prior knowledge about the study. 

It would be great if possible.
(I think that an appropriate validation (and error message) is also a practical solution though)

Reply:
Just an idea, but let me share another possible design.
```python
class Study:
  def save_to_sqlite(self, path):
      study = optuna.create_study(study_name=self.study_name, storage=f"sqlite:///{path}", directions=self.directions)
      ... copy trials, user-attrs, and system-attrs ...
```

To load the saved study, users can use the existing `optuna.study.load_study` function.

Reply:
Or,
```python
def copy_study(study_name, from_storage, to_storage):
  from_study = optuna.load_study(study_name=study_name, storage=from_storage)
  to_study = optuna.create_study(study_name=study_name, storage=to_storage directions=from_storage.directions)
      ... copy trials, user-attrs, and system-attrs ...
```

Reply:
Hm, I see. I was also imagining use cases where the storage information cannot be shared between the ones saving and loading. I'm not sure it's possible with this last design.

Reply:
If you specify an SQLite file path to `to_storage`, I think, you could handle the file in the same way with the file generated by `optuna.save_study(..)` in this PR.
(as a difference, users would have to know the target study name in addition to the file path)

Reply:
Right, thanks for the explanation. I was imagining always directly writing to the storage of the receiving user (in the case of e.g. MySQL).

Reply:
Leaving some conclusions from a discussion we had outside.

- Use SQLite as format. It will be less intrusive to the current API since we can reuse `load_study`. Additionally, behavior will be well defined in contrast to pickle
- Allow resolving conflicting study names at dump via argument. Could be a follow up

Possible follow-ups

- Functionality to convert between storages. E.g. via `copy_study` or arguments to `load_study`. It'll be useful for restoring studies and converting them to in-memory in favor of performance
- Allow omitting study name with `load_study` if there is only a single study in the storage

Reply:
I'll update the PR according to the items above.